File: cses/05_range_queries/1652_forest-queries/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần cho nhanh; mỗi hàng lưới là một token bytes.
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    STAR = 42  # mã ord('*'), so sánh trực tiếp trên bytes cho nhanh

    # Bảng cộng dồn hai chiều: P[i] là hàng prefix-sum độ dài n+1 (1-indexed).
    # P[0] toàn số 0 để làm biên cho công thức bù trừ.
    P = [[0] * (n + 1)]
    prev = P[0]
    for _ in range(n):
        row = data[idx]; idx += 1
        cur = [0] * (n + 1)
        s = 0  # cộng dồn số cây theo cột trong hàng hiện tại
        for j in range(n):
            if row[j] == STAR:
                s += 1
            # P[i][j] = P[i-1][j] + (cộng dồn theo cột trong hàng i)
            cur[j + 1] = prev[j + 1] + s
        P.append(cur)
        prev = cur

    out = []
    ap = out.append
    for _ in range(q):
        y1 = int(data[idx]); x1 = int(data[idx + 1])
        y2 = int(data[idx + 2]); x2 = int(data[idx + 3])
        idx += 4
        # Trả lời truy vấn O(1) bằng nguyên lý bù trừ (inclusion-exclusion).
        ap(P[y2][x2] - P[y1 - 1][x2] - P[y2][x1 - 1] + P[y1 - 1][x1 - 1])

    sys.stdout.buffer.write(('\n'.join(map(str, out)) + '\n').encode())


main()

File: cses/05_range_queries/1652_forest-queries/solution.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

static size_t to_index(int value) {
    return static_cast<size_t>(value);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;

    // Bảng cộng dồn hai chiều (2D prefix sum), 1-indexed; hàng/cột 0 để trống làm biên.
    vector<vector<int>> prefix(to_index(n + 1), vector<int>(to_index(n + 1), 0));
    for (int row = 1; row <= n; ++row) {
        string cells;
        cin >> cells;
        // row_sum = cộng dồn số cây trong hàng hiện tại; cộng thêm hàng phía trên
        // để có P[i][j] = P[i-1][j] + (cộng dồn theo cột trong hàng i).
        int row_sum = 0;
        for (int column = 1; column <= n; ++column) {
            row_sum += (cells[to_index(column - 1)] == '*');
            prefix[to_index(row)][to_index(column)] =
                prefix[to_index(row - 1)][to_index(column)] + row_sum;
        }
    }

    while (q--) {
        int y1, x1, y2, x2;
        cin >> y1 >> x1 >> y2 >> x2;
        // Trả lời truy vấn O(1) bằng nguyên lý bù trừ (inclusion-exclusion).
        const int answer = prefix[to_index(y2)][to_index(x2)]
                         - prefix[to_index(y1 - 1)][to_index(x2)]
                         - prefix[to_index(y2)][to_index(x1 - 1)]
                         + prefix[to_index(y1 - 1)][to_index(x1 - 1)];
        cout << answer << '\n';
    }

    return 0;
}
